src/evaluate/evaluate.ts

```diff
@@ -0,0 +1,360 @@
+import * as cdk from 'aws-cdk-lib';
+import { Construct, IConstruct } from 'constructs';
+import { SubFragment } from '../parser/private/sub';
+import {
+  assertBoolean,
+  assertList,
+  assertListOfForm,
+  assertNumber,
+  assertString,
+} from '../parser/private/types';
+import { assertExpression } from '../parser/template';
+import { ResourceOverride } from '../parser/template/overrides';
+import { ResourceTag } from '../parser/template/tags';
+import {
+  isCdkConstructExpression,
+  resolveResourceLike,
+  ResourceLike,
+} from '../type-resolution';
+import { TypedTemplateExpression } from '../type-resolution/expression';
+import { EvaluationContext } from './context';
+import { applyOverride } from './overrides';
+
+export class Evaluator {
+  private currentResource?: ResourceLike;
+
+  constructor(public readonly context: EvaluationContext) {}
+
+  public evaluateTemplate(block: <T>(fn: () => T) => T) {
+    this.context.template.resourceGraph().forEach((logicalId, resource) => {
+      block(() =>
+        this.evaluateResource(
+          logicalId,
+          resolveResourceLike(resource, logicalId, this.context.typeSystem)
```

Comment:
For everything else, type resolution is done by the parser. But resource-like is being resolved here, at the evaluation phase. Why is that?

Reply:
Good question. Short answer: No particular reason. We can move things around.

--- 

Grammar parsing, type resolution and evaluation are all distinct steps.
They all need to be initiated.

- Parsing is currently initiated here: https://github.com/cdklabs/decdk/blob/145238a53f1f9f8db5c1808500e6bac4c8a0d4a1/src/decdk.ts#L20
- Evaluation here: https://github.com/cdklabs/decdk/blob/82c309792aeb02330377c6e1382ca0eacd25eee9/src/declarative-stack.ts#L31

You are right to point out that the type resolution should be handled explicitly.

--- 

We will have to do something like this:
```
declare template: Template;
const typedTemplate: TypedTemplate = resolveTemplate(template);
new Evaluator(context).evaluateTemplate();
```